core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java

```diff
@@ -0,0 +1,879 @@
+import org.apache.carbondata.core.util.CarbonUtil;
+import org.apache.carbondata.core.util.DataFileFooterConverter;
+import org.apache.carbondata.core.util.path.CarbonTablePath;
+
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.fs.Path;
+import org.xerial.snappy.Snappy;
+
+/**
+ * Datamap implementation for blocklet.
+ */
+public class BlockDataMap extends CoarseGrainDataMap implements Serializable {
+
+  private static final LogService LOGGER =
+      LogServiceFactory.getLogService(BlockDataMap.class.getName());
+
+  protected static final long serialVersionUID = -2170289352240810993L;
+  /**
+   * for CACHE_LEVEL=BLOCK and legacy store default blocklet id will be -1
+   */
+  private static final short BLOCK_DEFAULT_BLOCKLET_ID = -1;
+
+  protected AbstractMemoryDMStore memoryDMStore;
+
+  protected AbstractMemoryDMStore taskSummaryDMStore;
+
+  // As it is a heavy object it is not recommended to serialize this object
+  protected transient SegmentProperties segmentProperties;
+
+  protected int[] columnCardinality;
+
+  protected long blockletSchemaTime;
+  /**
+   * flag to check for store from 1.1 or any prior version
+   */
+  protected boolean isLegacyStore;
+
+  @Override public void init(DataMapModel dataMapModel) throws IOException, MemoryException {
+    long startTime = System.currentTimeMillis();
+    assert (dataMapModel instanceof BlockletDataMapModel);
+    BlockletDataMapModel blockletDataMapInfo = (BlockletDataMapModel) dataMapModel;
+    DataFileFooterConverter fileFooterConverter = new DataFileFooterConverter();
+    List<DataFileFooter> indexInfo = fileFooterConverter
+        .getIndexInfo(blockletDataMapInfo.getFilePath(), blockletDataMapInfo.getFileData());
+    Path path = new Path(blockletDataMapInfo.getFilePath());
+    byte[] filePath = path.getParent().toString().getBytes(CarbonCommonConstants.DEFAULT_CHARSET);
+    byte[] fileName = path.getName().toString().getBytes(CarbonCommonConstants.DEFAULT_CHARSET);
+    byte[] segmentId =
+        blockletDataMapInfo.getSegmentId().getBytes(CarbonCommonConstants.DEFAULT_CHARSET);
+    byte[] schemaBinary = null;
+    if (!indexInfo.isEmpty()) {
+      DataFileFooter fileFooter = indexInfo.get(0);
+      // store for 1.1 or any prior version will not have any blocklet information in file footer
+      isLegacyStore = fileFooter.getBlockletList() == null;
+      // init segment properties and create schema
+      initSegmentProperties(fileFooter);
+      schemaBinary = convertSchemaToBinary(fileFooter.getColumnInTable());
+      createSchema(segmentProperties, blockletDataMapInfo.isAddToUnsafe(), true);
+      createSummarySchema(segmentProperties, schemaBinary, filePath, fileName, segmentId,
+          blockletDataMapInfo.isAddToUnsafe(), true);
+    }
+    // check for legacy store and load the metadata
+    DataMapRowImpl summaryRow = loadBlockMetadata(blockletDataMapInfo, indexInfo);
```

Comment:
Create protected methods for `createSchema`, `createSummarySchema` and `loadMetadata` and override the same in BloclkletDataMap